Docs: Add Ingress documentation

scholzj opened the pull request:
### Type of change

- Documentation

### Description

This PR adds the documentation to exposing Kafka using Kubernetes Ingress. This is related to the PR #1538 

### Checklist

- [x] Update documentation

tombentley commented on documentation/book/assembly-configuring-kafka-listeners.adoc:
```diff
@@ -11,10 +11,11 @@
 
 {ProductName} allows users to configure the listeners which will be enabled in Kafka brokers.
-Two types of listeners are supported:
+Three types of listeners are supported:
```
```suggestion
Three types of listener are supported:
```

tombentley commented on documentation/book/con-kafka-listeners.adoc:
```diff
@@ -148,7 +151,89 @@ listeners:
+
+{ProductName} is using the TLS passthrough feature of the {NginxIngressController}.
+Make sure TLS passthrough is enabled in your {NginxIngressController} deployment.
+For more information abotu enabling TLS passthrough see {NginxIngressControllerTLSPassthrough}.
```
```suggestion
For more information about enabling TLS passthrough see {NginxIngressControllerTLSPassthrough}.
```

PaulRMellor replied:
```suggestion
{ProductName} uses the TLS passthrough feature of the {NginxIngressController}.
```

tombentley commented on documentation/book/con-kafka-listeners.adoc:
```diff
@@ -148,7 +151,89 @@ listeners:
+Because it is using the TLS passthrough functionality, TLS encryption will be always used when exposing Kafka using `Ingress`.
+
+The Ingress controller does not assign any hostnames automatically.
+You have to specify the hostnames which should be used by the bootstrap and per-broker services in the `configuration` section.
```
Can we fully qualify the `configuration` section?

tombentley commented on documentation/book/con-kafka-listeners.adoc:
```diff
@@ -148,7 +151,89 @@ listeners:
+The Ingress controller does not assign any hostnames automatically.
+You have to specify the hostnames which should be used by the bootstrap and per-broker services in the `configuration` section.
+You also have to make sure that the hostnames resolve to the Ingress endpoints.
+{ProductName} will not perform any validation that the requested hosts are available and properly routed to the Ingress endpoints.
```
I suppose we _could_ do this, if we thought it was worth the effort.

scholzj replied:
That would be quite complicated ... we would need to setup our own backend, setup a test ingress for it and check that we can reach it. And that alone might still not work if there are some network restrictions, DNS limitations etc. So I would probably say that we could do it, but it would be very complicated.

tombentley commented on documentation/book/proc-accessing-kafka-using-ingress.adoc:
```diff
@@ -0,0 +1,64 @@
+[source,shell,subs=+quotes]
+oc apply -f _your-file_
+
+. Extract the public certificate of the broker certification authority
```
I _thought_ we had settled on "certificate authority", but maybe I'm wrong about that.

tombentley commented on documentation/book/proc-accessing-kafka-using-ingress.adoc:
```diff
@@ -0,0 +1,64 @@
+[source,shell]
+oc extract secret/_cluster-name_-cluster-ca-cert --keys=ca.crt --to=- > ca.crt
++
+. Use the extracted certificate in your Kafka client to configure TLS connection.
```
```suggestion
. Use the extracted certificate in your Kafka client to configure the TLS connection.
```

ncbaratta commented on documentation/book/con-kafka-listeners.adoc:
```diff
@@ -43,8 +43,11 @@ The external listener is used to connect to a Kafka cluster from outside of an {
+endif::Kubernetes[]
 
-.Exposing Kafka using {OpenShiftName} Routes
+=== Exposing Kafka using {OpenShiftName} Routes
```
Should Routes be `Routes` like it is in the following sentences?

ncbaratta commented on documentation/book/con-kafka-listeners.adoc:
```diff
@@ -85,7 +88,7 @@ listeners:
 For more information on using `Routes` to access Kafka, see xref:proc-accessing-kafka-using-routes-{context}[].
 
-.Exposing Kafka using loadbalancers
+=== Exposing Kafka using loadbalancers
```
Should loadbalancers be `loadbalancer` like it is in the following sentences?

scholzj commented:
@ncbaratta I think we were never consistent in the first place :-(. I just changed the titles because it is painful that we cannot link to these upstream. I didn't changed the spacing. 

I think (and updated the PR):
* Routes should be `Routes` because it refers to particular OpenShift API
* loadbalancers should stay loadbalancers because it refers to the pattern and not any particular API. (Where as on the line below it refers to the type as defined in the Kubernetes `Service` API)

But I can do it differently if desired.

ncbaratta commented:
I agree that we were never consistent, just wanted to try and make sure we get consistent before it becomes too hard to manage/clean up.  Your suggestion works for me! :)

scholzj commented:
Thanks @ncbaratta

PaulRMellor commented on documentation/book/con-kafka-listeners.adoc:
```diff
@@ -148,7 +151,89 @@ listeners:
+
+NOTE: External listeners using `Ingress` have been currently tested only with the {NginxIngressController}.
+
+{ProductName} is using the TLS passthrough feature of the {NginxIngressController}.
```
```suggestion
{ProductName} uses the TLS passthrough feature of the {NginxIngressController}.
```

PaulRMellor commented on documentation/book/proc-accessing-kafka-using-ingress.adoc:
```diff
@@ -0,0 +1,64 @@
+
+[id='proc-accessing-kafka-using-ingress-{context}']
+= Accessing Kafka using {KubernetesName} ingress
+
```
Shall we add an intro to the procedure? 'In this procedure, ...'

scholzj replied:
Done.

ppatierno commented on documentation/book/con-kafka-listeners.adoc:
```diff
@@ -148,7 +151,89 @@ listeners:
+{ProductName} uses the TLS passthrough feature of the {NginxIngressController}.
+Make sure TLS passthrough is enabled in your {NginxIngressController} deployment.
+For more information about enabling TLS passthrough see {NginxIngressControllerTLSPassthrough}.
+Because it is using the TLS passthrough functionality, TLS encryption will be always used when exposing Kafka using `Ingress`.
```
just a suggestion ... but maybe it's only me, I would add " ... TLS encryption on the Kafka side ..." in order to be more precise to explain that TLS passthrough means that Nginx will not handle TLS but in any case on Kafka you have to have it enable because the TLS handshake will happen there.

scholzj replied:
Does it really matter who actually handles it? You still have to use the TLS. The main point of this is to make it clear that you cannot disable TLS because of the TLS passthrough. Should it be maybe changed that way? 

_Because it is using the TLS passthrough functionality, TLS encryption cannot be disbled when exposing Kafka using `Ingress`._

ppatierno replied:
Sounds better ... even if I like to be explicit with the reader to drive him understanding that anyway the TLS will be handled by Kafka. But if you like this way, no need to change it's really a nit.
